### PR TITLE
Simplify classification caching checks

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Things were complicated here because we used to have a non-JTF-friendly blocking wait in one of our taggers.  That code is long gone.  So we can just do things the right way here.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs [hunk 1]
```diff
@@ -95,20 +98,6 @@ internal abstract class AbstractClassificationService : IClassificationService
...
-        {
-            var workspaceStatusService = document.Project.Solution.Services.GetRequiredService<IWorkspaceStatusService>();
-
-            // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
-            // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
-            // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
-            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
-            // fully loaded.
-            var isFullyLoadedTask = workspaceStatusService.IsFullyLoadedAsync(cancellationToken);
-            var isFullyLoaded = isFullyLoadedTask.IsCompleted && isFullyLoadedTask.GetAwaiter().GetResult();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs:108]: this was the wonky bit.  we're fine now, as the only place that does blocking in tagging is in the copy/paste tagger, and it uses JTF like so:

```c#
                    _owner._threadingContext.JoinableTaskFactory.Run(async () =>
                    {
                        var snapshotSpan = new DocumentSnapshotSpan(document, spanToTag);

                        // When copying/pasting, ensure we have classifications fully computed for the requested spans
                        // for both semantic classifications and embedded lang classifications.
                        await ProduceTagsAsync(context, snapshotSpan, classificationService, options, ClassificationType.Semantic, cancellationToken).ConfigureAwait(false);
                        await ProduceTagsAsync(context, snapshotSpan, classificationService, options, ClassificationType.EmbeddedLanguage, cancellationToken).ConfigureAwait(false);
                    });
```

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs [hunk 2]
```diff
@@ -246,7 +246,7 @@ private void ChangeProjectProperty<T>(ref T field, T newValue, Action<SolutionCh
                         // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
                         // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
                         // fully loaded.
-                        var isFullyLoadedTask = workspaceStatusService?.IsFullyLoadedAsync(CancellationToken.None);
+                        var isFullyLoadedTask = workspaceStatusService.IsFullyLoadedAsync(CancellationToken.None);
                         var isFullyLoaded = isFullyLoadedTask is { IsCompleted: true } && isFullyLoadedTask.GetAwaiter().GetResult();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:250]: this is still wonky.  but it's under a lock, so i'm not touching it.

File: src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 3]
```diff
@@ -155,30 +155,19 @@ public async Task WaitUntilFullyLoadedAsync(CancellationToken cancellationToken)
...
-                }
-
-                return !status.IsInProgress;
+                return status != null && !status.IsInProgress;
             }
 
             private async ValueTask<IVsOperationProgressStageStatusForSolutionLoad?> GetProgressStageStatusAsync(CancellationToken cancellationToken)
             {
                 // Workaround for lack of fast path in JoinAsync; avoid calling when already completed
                 // https://github.com/microsoft/vs-threading/pull/696
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: microsoft/vs-threading#696 was merged, so this no longer needs a workaround. Just use `JoinAsync` on both paths.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: @sharwell  do i need 'paths' at all at that point?  should just just be `await _progressStageStatus.JoinAsync(cancellationToken).ConfigureAwait(false)`?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: Right that's what I meant 😄

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: perfect.  thanks.

**Comment by CyrusNajmabadi** [comment]
@sharwell ptal.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide   @ToddGrun  ptal

File: src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 4]
```diff
@@ -155,31 +155,14 @@ public async Task WaitUntilFullyLoadedAsync(CancellationToken cancellationToken)
...
             public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
             {
                 var status = await GetProgressStageStatusAsync(cancellationToken).ConfigureAwait(false);
-                if (status == null)
-                {
-                    return false;
-                }
-
-                return !status.IsInProgress;
+                return status != null && !status.IsInProgress;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:161]: 📝 Combining the condition significantly reduced readability for me here

**Review by sharwell — APPROVED** [review]
Seems fine, but I wish dbbc90e53766e22841f1ada5b7b15b81c9621b88 wasn't merged into this branch

**Comment by CyrusNajmabadi** [comment]
https://github.com/orgs/dotnet/teams/roslyn-ide @ToddGrun ptal

**Pull request merged**